File: src/main/java/net/internetisalie/lunar/util/LuaFileUtil.java
package net.internetisalie.lunar.util;

import com.intellij.ide.plugins.PluginManagerCore;
import com.intellij.openapi.extensions.PluginId;
import com.intellij.openapi.util.Key;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileManager;
import net.internetisalie.lunar.LuaPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class LuaFileUtil {
    public static final Key<Boolean> PREDEFINED_KEY = Key.create("lua.lib.predefined");

    @NotNull
    public static String getPathToDisplay(final VirtualFile file) {
        if (file == null) {
            return "";
        }
        return FileUtil.toSystemDependentName(file.getPath());
    }

    @Nullable
    public static VirtualFile getPluginVirtualDirectory() {
        var descriptor = PluginManagerCore.getPlugin(PluginId.getId(LuaPlugin.ID));
        if (descriptor != null) {
            return VirtualFileManager.getInstance().findFileByNioPath(descriptor.getPluginPath());
        }

        return null;
    }

    @Nullable
    public static VirtualFile getPluginVirtualDirectoryChild(String ...args) {
        VirtualFile dir = LuaFileUtil.getPluginVirtualDirectory();
        for (String arg : args) {
            if (dir == null) break;
            dir = dir.findChild(arg);
        }
        return dir;
    }
}

File: src/main/kotlin/net/internetisalie/lunar/util/LuaFileUtil.kt
package net.internetisalie.lunar.util

import com.intellij.ide.plugins.PluginManagerCore.getPlugin
import com.intellij.openapi.extensions.PluginId
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.openapi.vfs.VirtualFileVisitor
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiManager
import net.internetisalie.lunar.LuaPlugin
import net.internetisalie.lunar.lang.LuaFileType

object LuaFileUtil {
    val pluginVirtualDirectory: VirtualFile?
        get() {
            val descriptor = getPlugin(PluginId.getId(LuaPlugin.ID)) ?: return null
            return VirtualFileManager.getInstance().findFileByNioPath(descriptor.pluginPath)
        }

    fun getPluginVirtualDirectoryChild(vararg args: String): VirtualFile? {
        var dir: VirtualFile? = pluginVirtualDirectory
        for (arg in args) {
            if (dir == null) break
            dir = dir.findChild(arg)
        }
        return dir
    }

    fun findLuaFilesInDir(dir: VirtualFile): List<VirtualFile> {
        val result = mutableListOf<VirtualFile>()
        VfsUtil.visitChildrenRecursively(dir, object : VirtualFileVisitor<VirtualFile>() {
            override fun visitFile(vf: VirtualFile): Boolean {
                if (vf.fileType == LuaFileType) {
                    result.add(vf)
                }
                return true
            }
        })
        return result
    }

    fun findPsiFiles(project: Project, virtualFiles : Collection<VirtualFile>) : List<PsiFile> {
        val psiManager = PsiManager.getInstance(project)
        return virtualFiles.mapNotNull { psiManager.findFile(it) }
    }
}
